keps/0036-kudo-operator-toggle-behavior.md

```diff
@@ -0,0 +1,147 @@
+---
+kep-number: 36
+short-desc: KudoOperator Toggling
+title: Parameter-controlled toggling of KudoOperator tasks
+authors:
+  - "@asekretenko"
+owners:
+  - "@asekretenko"
+editor: @asekretenko
+creation-date: 2021-04-30
+last-updated: 2021-04-30
+status: provisional
+see-also:
+  - KEP-23
+  - KEP-29
+---
+
+# Parameter-controlled toggling of KudoOperator tasks
+
+## Table of Contents
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Interface](#interface)
+     * [Validation](#validation)
+     * [Changing enablingParameter on upgrade](#changing-enablingparameter-on-upgrade)
+  * [Implementation Notes](#implementation-notes)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Drawbacks](#drawbacks)
+* [Alternatives](#alternatives)
+
+<!-- [Tools for generating]: https://github.com/ekalinin/github-markdown-toc -->
+
+## Summary
+
+This KEP aims to add an ability to switch an operator dependency on and off
+via a specified parameter of a parent operator. Namely, as a result of
+implementing this KEP, `KudoOperator` task definition will get an optional field
+containing a name of a boolean parameter (of the parent operator), the value
+of which will determine whether the child operator instance should exist.
+
+## Motivation
+From [#1775](https://github.com/kudobuilder/kudo/issues/1775):
+> Currently, KudoOperator tasks don't support toggle behavior which
+> significantly complicates the handling of the optional dependencies. A good
+> example could be Spark and History Server or Kafka and Schema Registry. The
+> only possible options to handle it today are either include the templates
+> as-is into the parent operator and control the contents at the template level
+> or to implement a job (via Toggle task) which deploys/removes the dependency
+> based on a parameter value. Also, cascading updates/upgrades become
+> complicated in these scenarios.
+
+### Goals
+
+* Add an ability to enforce existence/**non-existence** of the child
+  operator instance (and all its children) via a parent operator parameter,
+  i.e. to toggle the dependency.
```

Comment:
Could we please rephrase?   It is hard to understand and it is the goal.   Are we expecting KUDO to "enforce" something?  That isn't the mental model I formed to this point of reading.  I am expecting that a defined dependency is already installed and those I don't want KUDO to install it (as a dependency) of the parent kudo operator.  KUDO isn't "enforcing" anything at that point... it seems to be allowing a deviation from the current norm.

Reply:
@kensipe thanks, I'll rephrase this (and also other locations).

I should have been much more careful with the word "enforce". While **an execution of a task** (Apply/Delete/Toggle/KudoOperator) is enforcing existence/deletion of a resource (probably "is ensuring" would be a better word), neither tasks nor plans nor KUDO are really enforcing anything; such an enforcement is very far from the scope of this KEP.
Sorry for confusion.